firewood/src/storage/linear/mod.rs

```diff
@@ -68,47 +70,60 @@ mod filebacked;
 mod historical;
 mod layered;
 mod proposed;
-#[cfg(test)]
-mod tests;
 
 #[derive(Debug)]
-pub(super) struct LinearStore<S: ReadLinearStore> {
-    state: S,
-}
-
-/// All linearstores support reads
-pub(super) trait ReadLinearStore: Send + Sync + Debug {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error>;
-    fn size(&self) -> Result<u64, Error>;
+pub(super) enum ImmutableLinearStore {
+    Historical(Historical),
+    Proposed(Arc<Proposed<Immutable>>),
+    FileBacked(FileBacked),
+    Invalid, // TODO do we need this variant?
 }
 
-impl ReadLinearStore for Arc<dyn ReadLinearStore> {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
-        self.deref().stream_from(addr)
+impl ImmutableLinearStore {
+    pub fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.stream_from(addr),
+            ImmutableLinearStore::Proposed(proposed) => proposed.stream_from(addr),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.stream_from(addr),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
     }
 
-    fn size(&self) -> Result<u64, Error> {
-        self.deref().size()
+    pub fn size(&self) -> Result<u64, Error> {
+        match self {
+            ImmutableLinearStore::Historical(historical) => historical.size(),
+            ImmutableLinearStore::Proposed(proposed) => proposed.size(),
+            ImmutableLinearStore::FileBacked(filebacked) => filebacked.size(),
+            ImmutableLinearStore::Invalid => Err(std::io::ErrorKind::InvalidData.into()),
+        }
     }
 }
 
-/// Some linear stores support updates
-pub(super) trait WriteLinearStore: Debug {
-    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error>;
+#[derive(Debug)]
+pub(super) enum MutableLinearStore {
+    Proposed(Proposed<Mutable>),
```

Comment:
From what I've seen, you already have mutability built in. Let's say for a second that `Proposal` isn't generic over its mutability: `Proposal` (no `<T>`). 

Your write method requires mutable access, right? I can see here that the `Proposal` is owned and you can call `write` whenever you want. Your `ImmutableLinearStore` has an `Arc<Proposal>`. That's already immutable. You're not gaining anything (at least not that I can see) with the generic. 

Can you point me to the code where this is being used in a way that the type system can't already facilitate?